In [23]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [24]:
pd.set_option("display.max_columns",None)
pd.set_option("display.max_rows",None)

In [25]:
path=r"/content/data_reduce_cat_class.csv"
df=pd.read_csv(path)
df.shape

(7120, 108)

In [26]:
df.shape

(7120, 108)

In [27]:
df.head()

bath  balcony  price  total_sqft_int  bhk  price_per_sqft  \
0   3.0      2.0  150.0          1672.0    3     8971.291866   
1   3.0      3.0  149.0          1750.0    3     8514.285714   
2   3.0      2.0  150.0          1750.0    3     8571.428571   
3   2.0      2.0   40.0          1250.0    2     3200.000000   
4   2.0      2.0   83.0          1200.0    2     6916.666667   

   area_typeSuper built-up Area  area_typeBuilt-up Area  area_typePlot Area  \
0                             0                       0                   0   
1                             0                       0                   0   
2                             0                       0                   0   
3                             0                       0                   0   
4                             0                       0                   0   

   availability_Ready To Move  location_Whitefield  location_Sarjapur  Road  \
0                           1                    0                        0   
1                           1                    0                        0   
2                           1                    0                        0   
3                           1                    0                        0   
4                           1                    0                        0   

   location_Electronic City  location_Haralur Road  \
0                         0                      0   
1                         0                      0   
2                         0                      0   
3                         0                      0   
4                         0                      0   

   location_Raja Rajeshwari Nagar  location_Marathahalli  \
0                               0                      0   
1                               0                      0   
2                               0                      0   
3                               0                      0   
4                               0                      0   

   location_Hennur Road  location_Bannerghatta Road  location_Uttarahalli  \
0                     0                           0                     0   
1                     0                           0                     0   
2                     0                           0                     0   
3                     0                           0                     0   
4                     0                           0                     0   

   location_Thanisandra  location_Electronic City Phase II  location_Hebbal  \
0                     0                                  0                0   
1                     0                                  0                0   
2                     0                                  0                0   
3                     0                                  0                0   
4                     0                                  0                0   

   location_Yelahanka  location_7th Phase JP Nagar  location_Kanakpura Road  \
0                   0                            0                        0   
1                   0                            0                        0   
2                   0                            0                        0   
3                   0                            0                        0   
4                   0                            0                        0   

   location_KR Puram  location_Sarjapur  location_Rajaji Nagar  \
0                  0                  0                      0   
1                  0                  0                      0   
2                  0                  0                      0   
3                  0                  0                      0   
4                  0                  0                      0   

   location_Bellandur  location_Kasavanhalli  location_Begur Road  \
0                   0                      0                    0   
1                   0             

In [28]:
x=df.drop("price",axis=1)
y=df["price"]
print("shape of x=",x.shape)
print("shape of y=",y.shape)

shape of x= (7120, 107)
shape of y= (7120,)


In [29]:
#split dataset in train and test
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=51)
print("shape of x_train=",x_train.shape)
print("shape of y_train=",y_train.shape)
print("shape of x_test=",x_test.shape)
print("shape of y_test=",y_test.shape)

shape of x_train= (5696, 107)
shape of y_train= (5696,)
shape of x_test= (1424, 107)
shape of y_test= (1424,)


In [30]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
sc.fit(x_train)
x_train=sc.transform(x_train)
x_test=sc.transform(x_test)

In [31]:
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error
lr = LinearRegression()
lr_lasso = Lasso()
lr_ridge = Ridge()

In [32]:
def rmse(y_test, y_pred):
  return np.sqrt(mean_squared_error(y_test, y_pred))

In [33]:
lr.fit(x_train,y_train)
lr_score=lr.score(x_test,y_test)
lr_rmse=rmse(y_test,lr.predict(x_test))
lr_score,lr_rmse

(0.7917325013731912, 64.68930136543543)

In [34]:
lr_lasso.fit(x_train,y_train)
lr_lasso_score=lr_lasso.score(x_test,y_test)
lr_lasso_rmse=rmse(y_test,lr_lasso.predict(x_test))
lr_lasso_score,lr_lasso_rmse

(0.8038760088830822, 62.775050623930376)

In [35]:
lr_ridge.fit(x_train,y_train)
lr_ridge_score=lr_ridge.score(x_test,y_test)
lr_ridge_rmse=rmse(y_test,lr_ridge.predict(x_test))
lr_ridge_score,lr_ridge_rmse

(0.7919013501124277, 64.66307326506092)

In [36]:
from sklearn.svm import SVR
svr=SVR()
svr.fit(x_train,y_train)
svr_score=svr.score(x_test,y_test)
svr_rmse=rmse(y_test,svr.predict(x_test))
svr_score,svr_rmse

(0.21610213615728505, 125.50223235155066)

In [37]:
from sklearn.ensemble import RandomForestRegressor
rfr=RandomForestRegressor()
rfr.fit(x_train,y_train)
rfr_score=rfr.score(x_test,y_test)
rfr_rmse=rmse(y_test,rfr.predict(x_test))
rfr_score,rfr_rmse

(0.8918829069677491, 46.60890397363666)

In [38]:
import xgboost
xgb_reg=xgboost.XGBRegressor()
xgb_reg.fit(x_train,y_train)
xgb_reg_score=xgb_reg.score(x_test,y_test)
xgb_reg_rmse=rmse(y_test,xgb_reg.predict(x_test))
xgb_reg_score,xgb_reg_rmse

[04:47:46] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


(0.8795224594688463, 49.201093701747396)

In [39]:
print(pd.DataFrame([{'Model': 'Linear Regression','Score':lr_score, "RMSE":lr_rmse},
              {'Model': 'Lasso','Score':lr_lasso_score, "RMSE":lr_lasso_rmse},
              {'Model': 'Ridge','Score':lr_ridge_score, "RMSE":lr_ridge_rmse},
              {'Model': 'Support Vector Machine','Score':svr_score, "RMSE":svr_rmse},
              {'Model': 'Random Forest','Score':rfr_score, "RMSE":rfr_rmse},
              {'Model': 'XGBoost','Score':xgb_reg_score, "RMSE":xgb_reg_rmse}],
               columns=['Model','Score','RMSE']))

                    Model     Score        RMSE
0       Linear Regression  0.791733   64.689301
1                   Lasso  0.803876   62.775051
2                   Ridge  0.791901   64.663073
3  Support Vector Machine  0.216102  125.502232
4           Random Forest  0.891883   46.608904
5                 XGBoost  0.879522   49.201094


In [40]:
#cross validation
from sklearn.model_selection import KFold,cross_val_score
cvs = cross_val_score(xgb_reg,x_train,y_train,cv=10)
cvs,cvs.mean() 

[04:48:24] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
[04:48:26] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
[04:48:27] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
[04:48:28] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
[04:48:29] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
[04:48:30] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
[04:48:32] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
[04:48:33] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of re

(array([0.99380158, 0.99051863, 0.99023422, 0.98550236, 0.98553624,
        0.99224801, 0.98270072, 0.94283268, 0.99437309, 0.98687658]),
 0.9844624101293981)

In [20]:
cvs_rfr=cross_val_score(rfr,x_train,y_train,cv=10)
cvs_rfr,cvs_rfr.mean()

(array([0.99501918, 0.94535852, 0.99696932, 0.97265033, 0.96424898,
        0.90633637, 0.93137555, 0.90869196, 0.99625162, 0.99263704]),
 0.9609538868961097)

In [41]:
cvs_rfr2=cross_val_score(rfr,x_train,y_train,cv=10)
cvs_rfr2,cvs_rfr2.mean()

(array([0.99303543, 0.946166  , 0.99546808, 0.97517952, 0.965013  ,
        0.93472914, 0.93566916, 0.91383611, 0.99669426, 0.97990259]),
 0.9635693292850819)

In [ ]:
#Hyper parameter tunning
from sklearn.model_selection import GridSearchCV
from xgboost.sklearn import XGBRegressor
xgb1 = XGBRegressor()
parameters = {'learning_rate': [0.1,0.03, 0.05, 0.07],
              'min_child_weight': [1,3,5],
              'max_depth': [4, 6, 8],
              'gamma':[0,0.1,0.001,0.2],
              'subsample': [0.7,1,1.5],
              'colsample_bytree': [0.7,1,1.5],
              'objective':['reg:linear'],
              'n_estimators': [100,300,500]}
 
xgb_grid = GridSearchCV(xgb1,
                        parameters,
                        cv = 2,
                        n_jobs = -1,
                        verbose=True)
 
xgb_grid.fit(x_train,y_train)
 
print(xgb_grid.best_score_)
print(xgb_grid.best_params_)

Fitting 2 folds for each of 3888 candidates, totalling 7776 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  46 tasks      | elapsed:   50.7s
[Parallel(n_jobs=-1)]: Done 196 tasks      | elapsed:  5.0min
[Parallel(n_jobs=-1)]: Done 446 tasks      | elapsed: 11.5min
[Parallel(n_jobs=-1)]: Done 796 tasks      | elapsed: 20.9min
[Parallel(n_jobs=-1)]: Done 1246 tasks      | elapsed: 32.8min
[Parallel(n_jobs=-1)]: Done 1796 tasks      | elapsed: 48.1min
[Parallel(n_jobs=-1)]: Done 2446 tasks      | elapsed: 65.5min
[Parallel(n_jobs=-1)]: Done 3196 tasks      | elapsed: 90.6min
[Parallel(n_jobs=-1)]: Done 4046 tasks      | elapsed: 121.3min
[Parallel(n_jobs=-1)]: Done 4996 tasks      | elapsed: 154.6min
[Parallel(n_jobs=-1)]: Done 7776 out of 7776 | elapsed: 161.9min finished


[12:18:11] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
0.9534636832481846
{'colsample_bytree': 1, 'gamma': 0, 'learning_rate': 0.05, 'max_depth': 4, 'min_child_weight': 3, 'n_estimators': 500, 'objective': 'reg:linear', 'subsample': 0.7}


In [ ]:
xgb_tune=xgb_grid.estimator
xgb_tune.fit(x_train,y_train)
xgb_tune.score(x_test,y_test)

[12:57:26] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


0.8795224594688463

# Conclusion

1.From this project we can conclude that random forest regressor gives the 89% accuracy and the XGBoost gives the 87% accuracy
but the cross validation of the XGBoost gives the highest accuracy as compared to random forest regressor hence we consider
the XGBoost.

2.But using the hyperparameter it gives the 87% accuracy which is similar to XGBoost.

3.The price of houses are increases as we have the no of bathrooms,balcony etc.

# Thank you